sql: allow DISCARD in read-only transactions

michae2 opened the pull request:
To match PostgreSQL, allow DISCARD in read-only transactions.

Fixes: #124150

Release note (sql change): Fix a bug in which the DISCARD statement was disallowed with `default_transaction_read_only = on`.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/127298)
<!-- Reviewable:end -->

michae2 commented:
Hmm, this isn't quite right. `DISCARD` is schema-modifying (it drops temp tables). But PostgreSQL allows it anyway:

```
michae2=# CREATE TEMPORARY TABLE temp (i INT);
CREATE TABLE

michae2=# \dt temp
         List of relations
  Schema   | Name | Type  |  Owner
-----------+------+-------+---------
 pg_temp_3 | temp | table | michae2
(1 row)

michae2=# SET default_transaction_read_only = on;
SET

michae2=# DROP TABLE temp;
ERROR:  cannot execute DROP TABLE in a read-only transaction

michae2=# DISCARD temp;
DISCARD TEMP

michae2=# SHOW default_transaction_read_only;
 default_transaction_read_only
-------------------------------
 on
(1 row)

michae2=# \dt temp
Did not find any relation named "temp".
```

So I'm going to change this a bit, hold on one sec...

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127298)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`-- commits` line 10 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/127298#-O1xuW892h8WTkFCYpPw):*
`SetZoneConfig` is currently marked as `canModifySchema` yet it has `StatementReturnType` and `StatementType` as `RowsAffected` and `TypeDCL`, respectively, so it wouldn't get `true` for `CanModifySchema` anymore.


<!-- Sent from Reviewable.io -->

michae2 committed:
sql: allow DISCARD in read-only transactions

To match PostgreSQL, allow DISCARD in read-only transactions.

Fixes: #124150

Release note (sql change): Fix a bug in which the DISCARD statement was
disallowed with `default_transaction_read_only = on`.

michae2 commented:
Ok, fixed it in a different way. This should be RFAL.

yuzefovich reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Thanks!

> Reviewed 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127298#-:-O1yM1IQ2id0ZWPWFIEk:b-g2a5fb)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @rafiss)

___
*[`pkg/sql/logictest/testdata/logic_test/discard` line 165 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/127298#-O1yLVaCEPphSSAEhj8L:-O1yLVaCEPphSSAEhj8M:boarbm6) ([raw file](https://github.com/cockroachdb/cockroach/blob/2c8b473839aae479383d3abf987c2a9d54c3e598/pkg/sql/logictest/testdata/logic_test/discard#L165)):*
> ```
> 
> query T
> SET SEARCH_PATH = bar, public; SHOW SEARCH_PATH
> ```

nit: I think we usually use lower case for session variables, so `s/SEARCH_PATH/search_path/g` and `s/TIMEZONE/timezone/`.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
> Reviewed 2 of 3 files at r2, 1 of 1 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/127298)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2)



<!-- Sent from Reviewable.io -->

michae2 commented:
TFTRs!

bors r=yuzefovich,rafiss
